### PR TITLE
New calendar slot flow

### PR DESCRIPTION
**Pull request by doctorpod**

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -307,4 +316,4 @@ RUBY VERSION
    ruby 2.3.0p0
 
 BUNDLED WITH
-   1.14.3
+   1.13.6
```
Comment (StupidCodeFactory) [inline Gemfile.lock]: Could we keep this to the latest bundler version?

File: config/routes.rb [hunk 2]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  mount JasmineRails::Engine => '/specs' if defined?(JasmineRails)
```
Comment (StupidCodeFactory) [inline config/routes.rb]: might want to also check for ```Rails.env.test?```

File: app/models/slots_step.rb [hunk 3]
```diff
@@ -50,4 +66,17 @@ def slot_constraints
 
     @constraints ||= processor.booking_constraints.on_slots(show_live_slots)
   end
+
+  def next_slot_to_fill
+    return review_slot if review_slot.present?
+    return '0' if option_0.blank?
```
Comment (alan) [inline app/models/slots_step.rb]: Can we call `next_slot_to_add` here?

File: app/views/booking_requests/_booking_calendar.html.erb [hunk 4]
```diff
@@ -1,24 +1,27 @@
...
+        <% end %>
+      </tr>
+    </thead>
+    <tbody>
+      <tr><td id="errMsg" colspan="7">Javascript must be enabled</td></tr>
+    </tbody>
+  </table>
+
+  <div id="bn_prev-label" class="visuallyhidden">Go to previous month</div>
+  <div id="bn_next-label" class="visuallyhidden">Go to next month</div>
```
Comment (alan) [inline app/views/booking_requests/_booking_calendar.html.erb]: A few strings here need to be i18n

File: app/views/booking_requests/confirmation_step.html.erb [hunk 5]
```diff
@@ -48,47 +48,69 @@
...
-               url: booking_requests_path,
-               html: { class: 'js-SubmitOnce' }) do |f| %>
-            <%= render('hidden_prisoner_step') %>
-            <%= render('hidden_visitors_step') %>
-            <%= render('hidden_slots_step') %>
-            <%= hidden_field_tag('review_step', 'slots_step') %>
-            <%= f.submit(t('.change_visit_details'), class: 'button-link') %>
-          <% end %>
-      </th>
+      <th>Visit details</th>
```
Comment (alan) [inline app/views/booking_requests/confirmation_step.html.erb]: i18n

File: app/assets/javascripts/application.js [hunk 6]
```diff
@@ -27,11 +25,20 @@
...
-(function () {
+(function() {
   'use strict';
   delete moj.Modules.devs;
   var selectionButtons = new GOVUK.SelectionButtons("label input[type='radio'], label input[type='checkbox']");
   // console.log('moj', moj);
   moj.init();
-}());
+
+  $('#js-skipSlot').on('click', function(e) {
```
Comment (StupidCodeFactory) [inline app/assets/javascripts/application.js]: All Javascript we use follows as certain components pattern have a look at ```modules/*```

File: app/helpers/calendar_helper.rb [hunk 7]
```diff
@@ -44,4 +44,24 @@ def calendar_day(date, bookable)
...
+      displayed = s.iso8601
+      options = existing_selections.include?(displayed) ? chosen_options : {}
+      [format_slot_begin_time_for_public(s), displayed, options]
+    end
+  end
+
+  def chosen_options
+    options = {
+      'data-slot-chosen' => true,
+      'data-message' => 'Already chosen'
```
Comment (StupidCodeFactory) [inline app/helpers/calendar_helper.rb]: ```Already chosen``` needs to be translatable

File: app/views/booking_requests/_booking_calendar.html.erb [hunk 8]
```diff
@@ -1,24 +1,27 @@
...
+      <tr id="weekdays">
+        <% each_day_of_week do |day| %>
+          <th id="<%= I18n.l(day, format: :day_of_week) %>">
+            <abbr title="<%= I18n.l(day, format: :day_of_week) %>"><%= I18n.l(day, format: :day_of_week) %></abbr>
+          </th>
+        <% end %>
+      </tr>
+    </thead>
+    <tbody>
+      <tr><td id="errMsg" colspan="7">Javascript must be enabled</td></tr>
```
Comment (StupidCodeFactory) [inline app/views/booking_requests/_booking_calendar.html.erb]: I think this should be translatable as well

**Comment by doctorpod** [comment]
@alan I'm fixing the integration tests to work with new calendar flow - may make this a separate ticket...

**Pull request merged**